### PR TITLE
CRD implementation

### PR DESCRIPTION
**Pull request by kfswain**
This PR is rather large but much of this is generated code from these 2 tools:
- https://github.com/kubernetes/code-generator
- https://github.com/kubernetes-sigs/kubebuilder

The actual changes were:
- Wiring up code-generator to operate with kubebuilder on the `make generate` command
- Populating the API def with the spec as defined in the proposal #5 

Controller logic for the LLMServerPool is intended to be handled by gateway Implementations: #19 
And Controller logic for LLMService will be handled by the ext-proc deployment

### CODE REVIEW

File: PROJECT [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+projectName: api
+repo: inference.k8s.io
+resources:
+- api:
+    crdVersion: v1
+    namespaced: true
+  domain: inference.k8s.io
+  group: gateway
+  kind: LLMServerPool
+  path: inference.k8s.io/api/v1alpha1
```
Comment (ahg-g) [inline PROJECT:17]: sigs.k8s.io/llm-instance-gateway/api/inference/v1alpha2

File: PROJECT [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  version: v1alpha1
+- api:
```
Comment (ahg-g) [inline PROJECT:19]: ditto for this api too

File: api/README.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
+# api
```
Comment (ahg-g) [inline api/README.md:1]: We can reference the proposal here. Also, I think this readme should probably be one level up.

File: api/v1alpha1/groupversion_info.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+package v1alpha1
+
+import (
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/scheme"
+)
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "gateway.inference.k8s.io", Version: "v1alpha1"}
```
Comment (ahg-g) [inline api/v1alpha1/groupversion_info.go:29]: k8s.io is a reserved domain, we can't use it without an official api review from k8s.

btw, did we discuss the group name on the proposal? @robscott for some thoughts on the group name.

File: api/v1alpha1/llmserverpool_types.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// LLMServerPoolSpec defines the desired state of LLMServerPool
+type LLMServerPoolSpec struct {
+
+	// ModelServerSelector uses label selection to watch model server pods
+	// that should be included in the LLMServerPool. ModelServers should not
+	// be with any other Service or LLMServerPool, that behavior is not supported
+	// and will result in sub-optimal utilization.
+	ModelServerSelector map[string]string `json:"modelServerSelector,omitempty"`
```
Comment (ahg-g) [inline api/v1alpha1/llmserverpool_types.go:32]: Should we use a proper selector (i.e., metav1.LabelSelector)? @robscott ?

File: api/v1alpha1/llmserverpool_types.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+// LLMServerPoolStatus defines the observed state of LLMServerPool
+type LLMServerPoolStatus struct {
+
+	// Conditions track the state of the LLMServerPool.
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// ReadyServers are the number of available servers within the LLMServerPool.
+	ReadyServers int32 `json:"readyServers,omitempty"`
```
Comment (ahg-g) [inline api/v1alpha1/llmserverpool_types.go:42]: I am not sure if we will track this here. Service for example doesn't, and this is kinda equivalent to a service. I would start with Conditions only.

File: api/v1alpha1/llmservice_types.go [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// LLMService represents a set of LLM services that are multiplexed onto one
+// or more backend pools. This resource is managed by the "LLM Service Owner"
```
Comment (ahg-g) [inline api/v1alpha1/llmservice_types.go:29]: LLMServerPools

File: api/README.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+## Getting Started
+
+### Prerequisites
+- go version v1.22.0+
+- docker version 17.03+.
+- kubectl version v1.11.3+.
+- Access to a Kubernetes v1.11.3+ cluster.
+
+### To Deploy on the cluster
```
Comment (ahg-g) [inline api/README.md:15]: can we update this with the instructions to deploy the CRDs?

File: hack/boilerplate.go.txt [hunk 9]
```diff
@@ -0,0 +1,15 @@
+/*
+Copyright 2023.
```
Comment (ahg-g) [inline hack/boilerplate.go.txt:2]: 24

File: cmd/main.go [hunk 10]
```diff
@@ -0,0 +1,162 @@
+/*
```
Comment (ahg-g) [inline cmd/main.go:1]: Should we remove this?

File: api/README.md (hunk 3 above)
Comment (kfswain) [reply api/README.md:1]: Yeah, I just left this unmodified from how kubebuilder output it. I was going to merge this with our top-level readme later on, I suppose doing it this PR is as good as any

File: cmd/main.go (hunk 10 above)
Comment (kfswain) [reply cmd/main.go:1]: Yeah I was wondering that myself. I thought there may be a reasong that the file was made if there are no controllers. But if this is just manager scaffolding for eventual controllers, we can dump it.

File: PROJECT (hunk 1 above)
Comment (kfswain) [reply PROJECT]: Should we have inference in the path?

Comment (ahg-g) [reply PROJECT]: No, we can remove it. the path is indeed just `sigs.k8s.io/llm-instance-gateway/api/v1alpha1`

File: cmd/main.go (hunk 10 above)
Comment (ahg-g) [reply cmd/main.go:1]: Yeah, this is scaffolding for eventual controllers, we don't need it.

File: api/README.md (hunk 8 above)
Comment (kfswain) [reply api/README.md]: Added to the top-level README

File: api/v1alpha1/llmserverpool_types.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// LLMServerPoolSpec defines the desired state of LLMServerPool
+type LLMServerPoolSpec struct {
+
+	// ModelServerSelector uses label selection to watch model server pods
+	// that should be included in the LLMServerPool. ModelServers should not
+	// be with any other Service or LLMServerPool, that behavior is not supported
+	// and will result in sub-optimal utilization.
+	ModelServerSelector metav1.LabelSelector `json:"modelServerSelector,omitempty"`
```
[the remark this answers is not in the document]
Comment (ahg-g) [reply api/v1alpha1/llmserverpool_types.go:32]: The selector defines the pods that are running the servers, the expectation is that those pods [define a readiness probe and this readiness status is reflected on the pod status](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/); the expectation from the extension is to send traffic to the pods with status ready. 

> Also would it add any extra operation flexibility if we allowed more than one selectors?

The LabelSelector type is very flexible already, see https://github.com/kubernetes/kubernetes/blob/81ce66f059ec9c07cccf4069c8913e31959dea78/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1256

**Pull request merged**